pkg/firmware/utils.go

```diff
@@ -0,0 +1,284 @@
+		return "", "", fmt.Errorf("GetFirmwareVersionAndPSID(): firmware version (%v) or PSID (%v) is empty", firmwareVersion, PSID)
+	}
+
+	log.Log.V(2).Info("Firmware version and PSID found in .bin file", "version", firmwareVersion, "psid", PSID, "path", firmwareBinaryPath)
+
+	return firmwareVersion, PSID, nil
+}
+
+// CleanupDirectory deletes any file inside a root directory except for allowedSet. Empty directories are cleaned up as well at the end
+func (u utils) CleanupDirectory(root string, allowedSet map[string]struct{}) error {
+	log.Log.V(2).Info("FirmwareUtils.CleanupDirectory()", "root", root, "allowedSet", allowedSet)
+
+	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, walkErr error) error {
+		if walkErr != nil {
+			return walkErr
+		}
+
+		if d.IsDir() {
+			return nil
+		}
+
+		abs, err := filepath.Abs(path)
+		if err != nil {
+			return fmt.Errorf("failed to get absolute path of %q: %w", path, err)
+		}
+
+		if _, ok := allowedSet[abs]; !ok {
+			if err := os.Remove(abs); err != nil {
+				return fmt.Errorf("failed to remove %q: %w", abs, err)
+			}
+			log.Log.V(2).Info("deleted unaccounted file from cache dir", "path", abs, "cacheDir", root)
+		}
+		return nil
+	})
+
+	if err != nil {
+		err := fmt.Errorf("failed to walk directory for file removal: %w", err)
+		log.Log.Error(err, "failed to cleanup the cache directory", "cacheDir", root)
+	}
+
+	// After the unaccounted for files were deleted, clean up empty directories
+	if err := u.removeEmptyDirs(root); err != nil {
+		err := fmt.Errorf("failed to remove empty directories: %w", err)
+		log.Log.Error(err, "failed to cleanup the cache directory", "cacheDir", root)
+	}
+
+	return nil
+}
+
+// removeEmptyDirs recursively removes directories that are empty.
+// It does a post-order traversal: children first, then the parent.
+func (u utils) removeEmptyDirs(dir string) error {
+	entries, err := os.ReadDir(dir)
+	if err != nil {
+		return nil
+	}
+
+	// Recurse into subdirectories first
+	for _, entry := range entries {
+		if entry.IsDir() {
+			subDir := filepath.Join(dir, entry.Name())
+			if err := u.removeEmptyDirs(subDir); err != nil {
+				return err
+			}
+		}
+	}
+
+	// After processing children, check if 'dir' is now empty
+	// (Re-read directory to see if it has become empty)
+	entries, err = os.ReadDir(dir)
+	if err != nil {
+		return nil // If we can't re-read, just skip
+	}
+	if len(entries) == 0 && dir != "/" {
+		// Avoid removing root if you didn't intend to
+		if err := os.Remove(dir); err != nil {
+			return fmt.Errorf("failed to remove empty directory %q: %w", dir, err)
+		}
+		log.Log.V(2).Info("deleted empty directory", "dir", dir)
+	}
+
+	return nil
+}
+
+func newFirmwareUtils() FirmwareUtils {
+	return utils{execInterface: execUtils.New()}
+}
+
+// createDirIfNotExists checks if dir exists and creates it if not
+func createDirIfNotExists(path string) error {
+	if _, err := os.Stat(path); os.IsNotExist(err) {
```

Comment:
why do we need this? `MkdirAll` should work without this check according to 
https://pkg.go.dev/os#MkdirAll: " If path is already a directory, MkdirAll does nothing and returns nil"

Reply:
Thanks, deleted this function and replaced the calls with os.MkdirAll